clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java

```diff
@@ -3443,6 +3443,33 @@ public void testPreventMultiThread(GroupProtocol groupProtocol) throws Interrupt
         }
     }
 
+    @ParameterizedTest
+    @EnumSource(value = GroupProtocol.class)
+    public void testPollSendRequestForRebalance(GroupProtocol groupProtocol) throws InterruptedException {
+        ConsumerMetadata metadata = createMetadata(subscription);
+        MockClient client = new MockClient(time, metadata);
+        initMetadata(client, Collections.singletonMap(topic, 1));
+        Node node = metadata.fetch().nodes().get(0);
+
+        client.prepareResponseFrom(FindCoordinatorResponse.prepareResponse(Errors.NONE, groupId, node), node);
+        KafkaConsumer<String, String> consumer = newConsumer(groupProtocol, time, client, subscription, metadata,
+                assignor, true, groupInstanceId);
+        consumer.subscribe(singletonList(topic));
```

Comment:
Maybe we need to make `application thread`sleep here to make sure `network thread` does not send join request.
WDYT?

Reply:
Hi @TaiJuWu, `application` and `network` are different threads. I'm not sure why making `application` thread sleep can let `network` thread doesn't send join request. Also , this test would like to make sure `KafkaConsumer#subscribe` doesn't send join request, so we should not have other operations to make the result. Could you elaborate more about your idea? Thank you.

Reply:
Sorry for unclear.
Make `application` sleep here is just for making `network` have time to send `join request` without  poll in original code.
It can help to reproduce this bug easier and more stable, the test can pass sometime without it.

In order to make sure this behavior which we promise in docs, two operation should not too close in timeline.  Does it make sense?

Reply:
Hi @TaiJuWu, thanks for the suggestion. It's hard to say waiting how much time can make sure `KafkaConsumer#subscribe` doesn't send a join request. I add two unit tests to check the behavior.

In `ApplicationEventProcessorTest#testSubscriptionChangeEvent`, we check `ConsumerMembershipManager` doesn't call `maybeJoinGroup`.
In `ConsumerMembershipManagerTest#testOnSubscriptionUpdatedTransitionsToJoiningOnlyIfNotInGroup`, we check `ConsumerMembershipManager#onSubscriptionUpdated` keeps member state as `UNSUBSCRIBED` and calling `transitionToJoining` in `maybeJoinGroup`.

Reply:
Thanks for your explanation.